src/tiled/selectsametiletool.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+ * selectsametiletool.cpp
+ * Copyright 2009-2010, Jeff Bland <jksb@member.fsf.org>
+ * Copyright 2010, Thorbjørn Lindeijer <thorbjorn@lindeijer.nl>
+ * Copyright 2010, Jared Adams <jaxad0127@gmail.com>
+ * Copyright 2011, Stefan Beller <stefanbeller@googlemail.com>
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "selectsametiletool.h"
+
+#include "brushitem.h"
+#include "mapdocument.h"
+#include "changeselectedarea.h"
+
+#include <QApplication>
+
+using namespace Tiled;
+using namespace Tiled::Internal;
+
+SelectSameTileTool::SelectSameTileTool(QObject *parent)
+    : AbstractTileTool(tr("Select Same Tile"),
+                       QIcon(QLatin1String(
+                               ":images/22x22/stock-tool-select-same-tiles.png")),
+                       QKeySequence(tr("S")),
+                       parent)
+{
+}
+
+void SelectSameTileTool::tilePositionChanged(const QPoint &tilePos)
+{
+    // Make sure that a tile layer is selected and contains current tile pos.
+    TileLayer *tileLayer = currentTileLayer();
+    if (!tileLayer)
+        return;
+
+    QRegion resultRegion;
+    if (tileLayer->contains(tilePos)) {
+        const Cell matchCell = tileLayer->cellAt(tilePos);
+
+        for (int i = 0; i < tileLayer->height(); ++i) {
+            for (int j = 0; j < tileLayer->width(); ++j) {
+                if (tileLayer->cellAt(i, j) == matchCell)
+                    resultRegion += QRegion(i, j, 1, 1);
+            }
+        }
```

Comment:
Building up a `QRegion` as a series of small 1x1 rectangles gets potentially quite slow. For this reason, a much more efficient version exists as `TileLayer::region(Condition condition)`. Please change your code to use that one. You can use the condition `MatchesAnyTile` in `tilesetdock.cpp` as an example.